pkg/querier/queryrange/log_result_cache.go

```diff
@@ -0,0 +1,280 @@
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/common/model"
+	"github.com/weaveworks/common/httpgrpc"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/grafana/loki/pkg/loghttp"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/logqlmodel/stats"
+	"github.com/grafana/loki/pkg/querier/queryrange/queryrangebase"
+	"github.com/grafana/loki/pkg/storage/chunk/cache"
+	"github.com/grafana/loki/pkg/tenant"
+	"github.com/grafana/loki/pkg/util/validation"
+)
+
+// LogResultCacheMetrics is the metrics wrapper used in log result cache.
+type LogResultCacheMetrics struct {
+	CacheHit  prometheus.Counter
+	CacheMiss prometheus.Counter
+}
+
+// NewLogResultCacheMetrics creates metrics to be used in log result cache.
+func NewLogResultCacheMetrics(registerer prometheus.Registerer) *LogResultCacheMetrics {
+	return &LogResultCacheMetrics{
+		CacheHit: promauto.With(registerer).NewCounter(prometheus.CounterOpts{
+			Namespace: "loki",
+			Name:      "query_frontend_log_result_cache_hit_total",
+		}),
+		CacheMiss: promauto.With(registerer).NewCounter(prometheus.CounterOpts{
+			Namespace: "loki",
+			Name:      "query_frontend_log_result_cache_miss_total",
+		}),
+	}
+}
+
+// NewLogResultCache creates a new log result cache middleware.
+// Currently it only caches empty filter queries, this is because those are usually easily and freely cacheable.
+// Log hits are difficult to handle because of the limit query parameter and the size of the response.
+// In the future it could be extended to cache non-empty query results.
+// see https://docs.google.com/document/d/1_mACOpxdWZ5K0cIedaja5gzMbv-m0lUVazqZd2O4mEU/edit
+func NewLogResultCache(logger log.Logger, limits Limits, cache cache.Cache, shouldCache queryrangebase.ShouldCacheFn, metrics *LogResultCacheMetrics) queryrangebase.Middleware {
+	if metrics == nil {
+		metrics = NewLogResultCacheMetrics(nil)
+	}
+	return queryrangebase.MiddlewareFunc(func(next queryrangebase.Handler) queryrangebase.Handler {
+		return &logResultCache{
+			next:        next,
+			limits:      limits,
+			cache:       cache,
+			logger:      logger,
+			shouldCache: shouldCache,
+			metrics:     metrics,
+		}
+	})
+}
+
+type logResultCache struct {
+	next        queryrangebase.Handler
+	limits      Limits
+	cache       cache.Cache
+	shouldCache queryrangebase.ShouldCacheFn
+
+	metrics *LogResultCacheMetrics
+	logger  log.Logger
+}
+
+func (l *logResultCache) Do(ctx context.Context, req queryrangebase.Request) (queryrangebase.Response, error) {
+	tenantIDs, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+	}
+
+	if l.shouldCache != nil && !l.shouldCache(req) {
+		return l.next.Do(ctx, req)
+	}
+
+	maxCacheFreshness := validation.MaxDurationPerTenant(tenantIDs, l.limits.MaxCacheFreshness)
+	maxCacheTime := int64(model.Now().Add(-maxCacheFreshness))
+	if req.GetStart() > maxCacheTime {
```

Comment:
What about when `req.End > maxCacheTime`? I figure we should avoid caching requests that overlap with the freshness requirement, rather than those that just start in that period. Otherwise, we could cache a request that has no data then receive a push with data.

Reply:
Just requiring the end would do then because the start is always before.